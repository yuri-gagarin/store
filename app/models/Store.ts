import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;