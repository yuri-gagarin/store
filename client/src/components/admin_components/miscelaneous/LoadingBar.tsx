import React from "react";
import { Dimmer, Loader, Segment } from "semantic-ui-react";

interface Props {
  message?: string;
}
const LoadingBar: React.FC<Props> = ({ message }): JSX.Element => {

  return (
    <div style={{ height: "50px", width: "100%" }}>
      <Segment>
        <Dimmer active>
          <Loader>{message ? message : "Loading"}</Loader>
        </Dimmer>
      </Segment>
    </div>
  );
};

export default LoadingBar;
