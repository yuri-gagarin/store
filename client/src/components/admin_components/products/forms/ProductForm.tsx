import React, { useState, useEffect, useRef, memo }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";
// type definitions//
import { ProductFormState } from "../type_definitions/productTypes";

interface Props {
  name: string;
  price: string;
  description: string;
  details: string;
  newForm: boolean;
  handleCreateProduct(data: ProductFormState): void;
  handleUpdateProduct(data: ProductFormState): void;
}
class ProductForm extends React.Component<Props, ProductFormState> {
  constructor(props: Props) {
    super(props)
    // this.productFormRef = useRef<HTMLDivElement>(document.createElement("div"));
    this.state = {
      name: "",
      price: "",
      description: "",
      details: "",
    }

  }
  // local form state //
  // form ref //
  
  // form state changes listeners //
  handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    this.setState({
      ...this.state,
      name: e.target.value
    });
  };
  handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      price: e.target.value
    });
  };
  hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      description: e.target.value
    });
  };
  handleDetailsChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      details: e.target.value
    });
  };
  // submit action //
  handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.props.newForm) {
      this.props.handleCreateProduct(this.state);
    } else {
      this.props.handleUpdateProduct(this.state);
    }
  };
  componentDidMount() {

  }
  shouldComponentUpdate(nextProps: Props, nextState: ProductFormState) {
    console.log(nextProps)
    const { name: newName } = nextProps;
    if (newName === this.props.name) {
      console.log("shouldnt");
      return false
    }
    return true;
  }

  // lifecycle hooks //
  /*
  useEffect(() => {
    if (productFormRef.current) {
      const elem = productFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [productFormRef]);
  */
  render() {
    const { name, price, description, details } = this.state;
    const { newForm } = this.props;
    return (
      <div className="productFormDiv">
        <Form id="productForm">
          <Form.Field>
            <label>Product name</label>
            <input 
              id="adminProductFormNameInput"
              onChange={this.handleNameChange} 
              placeholder="Product name here ..." 
              value={name}
            />
          </Form.Field>
          <Form.Field>
            <label>Product price</label>
            <input 
              id="adminProductFormPriceInput"
              onChange={this.handlePriceChange} 
              placeholder="Product price here..." 
              value={price}
            />
          </Form.Field>
          <Form.Field
            id='adminProductFormDescInput'
            control={TextArea}
            label='Product Description'
            onChange={this.hadnleDescriptionChange}
            placeholder='Product description here...'
            value={description}
           />
          <Form.Field
            id='adminProductFormDetailsInput'
            control={TextArea}
            label='Product Details'
            onChange={this.handleDetailsChange}
            placeholder='Product details here...'
            value={details}
          />
        </Form>
        {
          newForm 
            ? <Button 
                id="adminProductFormCreate" 
                type="button"
                onClick={this.handleSubmit} 
                content= "Create  New Product" 
              />
            : <Button 
              id="adminProductFormUpdate" 
              type='button'
              onClick={this.handleSubmit} 
              content="Update Product" 
            />
        }  
      </div>
    );
  }
}

export default ProductForm;