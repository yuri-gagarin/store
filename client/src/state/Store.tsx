import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

