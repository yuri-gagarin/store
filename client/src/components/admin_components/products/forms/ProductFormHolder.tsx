import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// client routing //
import { RouteComponentProps, withRouter } from "react-router-dom";
// css imports //
import "./css/productFormHolder.css";
// additional components //
import ProductForm from "./ProductForm";
import ProductImgPreviewHolder from "../image_preview/ProductImgPreviewHolder";
import LoadingBar from "../../miscelaneous/LoadingBar";
import ProductImgUplForm from "./ProductImgUplForm";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createProduct, editProduct } from "../actions/APIProductActions";
import { closeProductForm, openProductForm } from "../actions/UIProductActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { FormState } from "./ProductForm";


interface Props extends RouteComponentProps {
 
}

type ProductData = {
  name: string;
  description: string;
  price: string;
  images: IProductImgData[];
}

const ProductFormHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state,dispatch } = useContext(Store);
  const { loading, currentProductData, productFormOpen, error } = state.productState;
  const { name, description, price, createdAt, editedAt } = currentProductData;
  // local component state //
  // const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);

  // ProductForm toggle //
  const toggleProductForm = () => {
    productFormOpen ? closeProductForm(dispatch) : openProductForm(dispatch);
  };
  // API call handlers CREATE - EDIT //
  const handleCreateProduct = ({ name, price, description }: FormState): void => {
    const productData: ProductData = {
      name,
      description,
      price,
      images: currentProductData.images
    };
    createProduct(productData, dispatch)
      .then((_) => {
        // product created //
        closeProductForm(dispatch);
      })
      .catch((_) => {
        // handle error or show modal //
      });
  };
  const handleUpdateProduct = ({ name, description, price }: FormState): void => {
    const productParams: ProductData = {
      name, description, price, images: currentProductData.images
    };
    editProduct(currentProductData._id, productParams, dispatch, state)
      .then((_) => {
        closeProductForm(dispatch)
      })
      .catch((_) => {
        // handle error show modal //
      });
  };
  // lifecycle hooks //
  useEffect(() => {
    if (!productFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ productFormOpen ]);
  useEffect(() => {
    checkSetValues(currentProductData) ? setNewForm(false) : setNewForm(true);
  }, [ currentProductData ]);
  // component return //
  return (
    <div id="productFormHolder">
      { loading ? <LoadingBar /> : null }
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row id={"productFormHolderDetails"}> 
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="productFormHolderDetailsItem">
                  <h3>Product name:</h3>
                  <p>{name}</p>
                </div>
                <div className="productFormHolderDetailsItem">
                  <h3>Product price:</h3>
                  <p>{price}</p>
                </div>
                <div className="productFormHolderDetailsItem">
                  <h3>Product description:</h3>
                  <p>{description}</p>
                </div>
                <div className="productFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <ProductImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <ProductImgUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  
            id="productFormToggleBtn"
            onClick={toggleProductForm} 
            content={ !productFormOpen ? "Open Form" : "Close Form"}
          />
          {
            productFormOpen ? 
              <ProductForm 
                name={name} 
                description={description} 
                price={price}
                handleCreateProduct={handleCreateProduct}
                handleUpdateProduct={handleUpdateProduct}
              /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};
// export without router for tests //
export { ProductFormHolder };
// default export //
export default withRouter(ProductFormHolder);

