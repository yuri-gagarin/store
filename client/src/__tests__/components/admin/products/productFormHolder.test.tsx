import React from "react"
import { Button } from "semantic-ui-react";
// testing utils
import { mount, ReactWrapper } from "enzyme";
import moxios from "moxios";
import { act } from "react-dom/test-utils";
// client routing //
import { MemoryRouter as Router } from "react-router-dom";
// component imports //
import ProductFormHolder from "../../../../components/admin_components/products/forms/ProductFormHolder";
import ProductForm from "../../../../components/admin_components/products/forms/ProductForm";
import ProductImageUplForm from "../../../../components/admin_components/products/forms/ProductImgUplForm";
import ProductImgPreviewHolder from "../../../../components/admin_components/products/image_preview/ProductImgPreviewHolder";
import ProductImgPreviewThumb from "../../../../components/admin_components/products/image_preview/ProductImgThumb";
import LoadingBar from "../../../../components/admin_components/miscelaneous/LoadingBar";
// state React.Context //
import { IGlobalAppState, TestStateProvider } from "../../../../state/Store";
// helpers //
import { createMockProducts, setMockProductState } from "../../../../test_helpers/productHelpers";
import { generateCleanState } from "../../../../test_helpers/miscHelpers";

describe("ProductFormHolder Component tests", () => {
  let wrapper: ReactWrapper; 
  describe("Default Form Holder state",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(
        <Router keyLength={0}>
          <ProductFormHolder />
        </Router>
      );
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find(ProductFormHolder)).toMatchSnapshot();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(ProductForm);
      expect(form.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(Button);
      expect(toggleButton.length).toEqual(1);
    });

  });
  
  // TEST Form Holder state OPEN - NO Current Product Data //
  describe("Form Holder state OPEN - NO Current Product Data",  () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider>
            <ProductFormHolder />
          </TestStateProvider>
        </Router>
      );
    });
    it("Should render a Form toggle Button", () => {
      const toggleButton = wrapper.find(ProductFormHolder).render().find('#productFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should Properly Mount Form Holder, respond to '#productToggleBtn' click", () => {
      const toggleButton = wrapper.find("#productFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // open button clicked //
      //wrapper.update()
      expect(wrapper.find(ProductFormHolder)).toMatchSnapshot();
    });
  
    it("Should have a Form Create Button", () => {
      const toggleButton = wrapper.find(ProductFormHolder).render().find('#adminProductFormCreate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered after toggle button", () => {
      const form = wrapper.find(ProductForm);
      expect(form.length).toEqual(1);
    });
    it("Should NOT render '#productFormHolderDetails", () => {
      const details = wrapper.find(ProductFormHolder).render().find("#productFormHolderDetails");
      expect(details.length).toEqual(0);
    })
    it("Should NOT have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(ProductImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(0);
    });
    it("Should NOT have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ProductImageUplForm);
      expect(imgUploadForm.length).toEqual(0);
    });
  });
  // END Form Holder state OPEN - NO Current Product Data //
  // TEST Form Holder state OPEN - WITH Current Product Data - NO IMAGES //
  describe("Form Holder state OPEN - WITH Current Product Data - NO IMAGES",  () => {
    let wrapper: ReactWrapper; let state: IGlobalAppState;
    let mockProduct: IProductData;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = generateCleanState();
      mockProduct = {
        _id: "1",
        price: "100",
        name: "name",
        description: "description",
        details: "details",
        images: [],
        createdAt: "1111"
      };
      state.productState.currentProductData = mockProduct;
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider mockState={state}>
            <ProductFormHolder />
          </TestStateProvider>
        </Router>
      );
      wrapper.update();
    });

    it("Should Properly render Form Holder", () => {
      const formHolder = wrapper.find(ProductFormHolder).find("#productFormHolder");
      expect(formHolder.length).toEqual(1);
    });
    it("Should render a Form toggle Button", () => {
      const toggleButton = wrapper.find(ProductFormHolder).render().find('#productFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should properly render ProductForm", () => {
      const toggleButton = wrapper.find("#productFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // assert form open //
      const form = wrapper.find(ProductForm);
      expect(form.length).toEqual(1);
    });
    it("Should render a Form Update Button", () => {
      const toggleButton = wrapper.find(ProductFormHolder).render().find('#adminProductFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should render '#productFormHolderDetails", () => {
      const details = wrapper.find(ProductFormHolder).render().find("#productFormHolderDetails");
      expect(details.length).toEqual(1);
    });
    it("Should render correct values in '#productFormHolderDetails", () => {
      const detailsHolders = wrapper.find(ProductFormHolder).find(".productFormHolderDetailsItem");
      expect(detailsHolders.at(0).find("p").render().text()).toEqual(mockProduct.name);
      expect(detailsHolders.at(1).find("p").render().text()).toEqual(mockProduct.price);
      expect(detailsHolders.at(2).find("p").render().text()).toEqual(mockProduct.description);
    })
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(ProductImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should NOT render any preview images", () => {
      const previewThumb = wrapper.find(ProductImgPreviewThumb);
      expect(previewThumb.length).toEqual(0);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ProductImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  // END Form Holder state OPEN - WITH Current Product Data - NO IMAGES //
  // TEST Form Holder state OPEN - WITH Current Product Data - WITH IMAGES //
  describe("Form Holder state OPEN - WITH Current Product Data - WITH IMAGES",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;
    let mockProduct: IProductData;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      // mock data //
      state = generateCleanState();
      mockProduct = {
        _id: "1",
        price: "100",
        name: "name",
        description: "description",
        details: "details",
        images: [
          {
            _id: "1",
            description: "desc",
            url: "url",
            absolutePath: "path",
            fileName: "img",
            imagePath: "imgPath",
            createdAt: "now"
          },
          {
            _id: "2",
            description: "desc",
            url: "url",
            absolutePath: "path",
            fileName: "img",
            imagePath: "imgPath",
            createdAt: "now"
          },
        ],
        createdAt: "1111"
      };
      state.productState.currentProductData = mockProduct;
      // mount component //
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider mockState={state}>
            <ProductFormHolder />
          </TestStateProvider>
        </Router>
      );
    });

    it("Should Properly render FormHolder", () => {
      const formHolder = wrapper.find(ProductFormHolder).find("#productFormHolder");
      expect(formHolder.length).toEqual(1);
    });
    it("Should render a Form toggle Button", () => {
      const toggleButton = wrapper.find(ProductFormHolder).render().find('#productFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const toggleButton = wrapper.find("#productFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // assert proper Form rendering //
      const form = wrapper.find(ProductForm);
      expect(form.length).toEqual(1);
    });
    it("Should render a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminProductFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should redner 'ProductImgPreview Holder'", () => {
      const imgPreviewHolder = wrapper.find(ProductImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should render a correct number of preview images", () => {
      const previewThumb = wrapper.find(ProductImgPreviewThumb);
      const numberOfImages = state.productState.currentProductData.images.length;
      expect(previewThumb.length).toEqual(numberOfImages);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ProductImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  // END Form Holder state OPEN - WITH Current Product Data - WITH IMAGES //
  // TEST Form Holder state OPEN - MOCK Submit action //
  describe("Form Holder state OPEN - MOCK Submit action",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;

    beforeAll( async () => {
      window.scrollTo = jest.fn();
      // mount and wait //
      wrapper = mount(
        <Router keyLength={0} initialEntries={["/admin/products/create"]} >
          <TestStateProvider>
            <ProductFormHolder />
          </TestStateProvider>
        </Router>
      );
    });

    it("Should have a submit button", () => {
      wrapper.update();
      wrapper.find("#productFormToggleBtn").at(0).simulate("click").update();
      const adminProductFormCreate = wrapper.find("#adminProductFormCreate").at(0);
      expect(adminProductFormCreate.length).toEqual(1)
    });
    it("Should handle the 'handleCreateProductAction, show 'LoadingBar' Component", async () => {
      moxios.install();
      await act( async () => {
        moxios.stubRequest("/api/products/create", {
          status: 200,
          response: {
            responseMsg: "All Good",
            newProduct: createMockProducts(1)[0]
          }
        });
        const adminProductFormCreate = wrapper.find("#adminProductFormCreate").at(0);
        adminProductFormCreate.simulate("click");
        //expect(wrapper.find(LoadingBar).length).toEqual(1);
      });
      // expect(sinon.spy(createProduct)).toHaveBeenCalled()
      wrapper.update();
    });
    it("Should NOT show the 'LoadingBar' Component after successful API call", () => {
      expect(wrapper.find(LoadingBar).length).toEqual(0);
    });
    it("Should NOT show the 'ProductForm' Component after successful API call", () => {
      expect(wrapper.find(ProductForm).length).toEqual(0);
    });
    // END Form Holder state OPEN - MOCK Submit action //
  });
  
});