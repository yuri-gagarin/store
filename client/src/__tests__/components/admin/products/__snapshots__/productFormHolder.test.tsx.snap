// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductFormHolder Component tests Default Form Holder state Should Properly Mount Form Holder 1`] = `
<withRouter(ProductFormHolder)>
  <ProductFormHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <div
      id="productFormHolder"
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            computer={14}
            mobile={16}
            tablet={15}
          >
            <div
              className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
            >
              <Button
                as="button"
                content="Open Form"
                id="productFormToggleBtn"
                onClick={[Function]}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="productFormToggleBtn"
                      >
                        Open Form
                      </button>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <button
                          class="ui button"
                          id="productFormToggleBtn"
                        >
                          Open Form
                        </button>,
                      }
                    }
                  >
                    <button
                      className="ui button"
                      id="productFormToggleBtn"
                      onClick={[Function]}
                    >
                      Open Form
                    </button>
                  </RefFindNode>
                </Ref>
              </Button>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </ProductFormHolder>
</withRouter(ProductFormHolder)>
`;

exports[`ProductFormHolder Component tests Form Holder state OPEN - NO Current Product Data Should Properly Mount Form Holder, respond to '#productToggleBtn' click 1`] = `
<withRouter(ProductFormHolder)>
  <ProductFormHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <div
      id="productFormHolder"
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            computer={14}
            mobile={16}
            tablet={15}
          >
            <div
              className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
            >
              <Button
                as="button"
                content="Close Form"
                id="productFormToggleBtn"
                onClick={[Function]}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="productFormToggleBtn"
                      >
                        Close Form
                      </button>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <button
                          class="ui button"
                          id="productFormToggleBtn"
                        >
                          Close Form
                        </button>,
                      }
                    }
                  >
                    <button
                      className="ui button"
                      id="productFormToggleBtn"
                      onClick={[Function]}
                    >
                      Close Form
                    </button>
                  </RefFindNode>
                </Ref>
              </Button>
              <ProductForm
                description=""
                details=""
                handleCreateProduct={[Function]}
                handleUpdateProduct={[Function]}
                name=""
                newForm={true}
                price=""
              >
                <div
                  className="productFormDiv"
                >
                  <Form
                    as="form"
                    id="productForm"
                  >
                    <form
                      className="ui form"
                      id="productForm"
                      onSubmit={[Function]}
                    >
                      <FormField>
                        <div
                          className="field"
                        >
                          <label>
                            Product name
                          </label>
                          <input
                            id="productFormNameInput"
                            onChange={[Function]}
                            placeholder="Product name here ..."
                            value=""
                          />
                        </div>
                      </FormField>
                      <FormField>
                        <div
                          className="field"
                        >
                          <label>
                            Product price
                          </label>
                          <input
                            id="productFormPriceInput"
                            onChange={[Function]}
                            placeholder="Product price here..."
                            value=""
                          />
                        </div>
                      </FormField>
                      <FormField
                        control={[Function]}
                        id="productFormDescInput"
                        label="Product Description"
                        onChange={[Function]}
                        placeholder="Product description here..."
                        value=""
                      >
                        <div
                          className="field"
                        >
                          <label
                            htmlFor="productFormDescInput"
                          >
                            Product Description
                          </label>
                          <TextArea
                            aria-describedby={null}
                            as="textarea"
                            id="productFormDescInput"
                            onChange={[Function]}
                            placeholder="Product description here..."
                            rows={3}
                            value=""
                          >
                            <Ref
                              innerRef={
                                Object {
                                  "current": <textarea
                                    id="productFormDescInput"
                                    placeholder="Product description here..."
                                    rows="3"
                                  />,
                                }
                              }
                            >
                              <RefFindNode
                                innerRef={
                                  Object {
                                    "current": <textarea
                                      id="productFormDescInput"
                                      placeholder="Product description here..."
                                      rows="3"
                                    />,
                                  }
                                }
                              >
                                <textarea
                                  aria-describedby={null}
                                  id="productFormDescInput"
                                  onChange={[Function]}
                                  onInput={[Function]}
                                  placeholder="Product description here..."
                                  rows={3}
                                  value=""
                                />
                              </RefFindNode>
                            </Ref>
                          </TextArea>
                        </div>
                      </FormField>
                      <FormField
                        control={[Function]}
                        id="productFormDetailsInput"
                        label="Product Details"
                        onChange={[Function]}
                        placeholder="Product details here..."
                        value=""
                      >
                        <div
                          className="field"
                        >
                          <label
                            htmlFor="productFormDetailsInput"
                          >
                            Product Details
                          </label>
                          <TextArea
                            aria-describedby={null}
                            as="textarea"
                            id="productFormDetailsInput"
                            onChange={[Function]}
                            placeholder="Product details here..."
                            rows={3}
                            value=""
                          >
                            <Ref
                              innerRef={
                                Object {
                                  "current": <textarea
                                    id="productFormDetailsInput"
                                    placeholder="Product details here..."
                                    rows="3"
                                  />,
                                }
                              }
                            >
                              <RefFindNode
                                innerRef={
                                  Object {
                                    "current": <textarea
                                      id="productFormDetailsInput"
                                      placeholder="Product details here..."
                                      rows="3"
                                    />,
                                  }
                                }
                              >
                                <textarea
                                  aria-describedby={null}
                                  id="productFormDetailsInput"
                                  onChange={[Function]}
                                  onInput={[Function]}
                                  placeholder="Product details here..."
                                  rows={3}
                                  value=""
                                />
                              </RefFindNode>
                            </Ref>
                          </TextArea>
                        </div>
                      </FormField>
                      <Button
                        as="button"
                        content="Create  New Product"
                        id="adminProductFormCreate"
                        onClick={[Function]}
                        type="submit"
                      >
                        <Ref
                          innerRef={
                            Object {
                              "current": <button
                                class="ui button"
                                id="adminProductFormCreate"
                                type="submit"
                              >
                                Create  New Product
                              </button>,
                            }
                          }
                        >
                          <RefFindNode
                            innerRef={
                              Object {
                                "current": <button
                                  class="ui button"
                                  id="adminProductFormCreate"
                                  type="submit"
                                >
                                  Create  New Product
                                </button>,
                              }
                            }
                          >
                            <button
                              className="ui button"
                              id="adminProductFormCreate"
                              onClick={[Function]}
                              type="submit"
                            >
                              Create  New Product
                            </button>
                          </RefFindNode>
                        </Ref>
                      </Button>
                    </form>
                  </Form>
                </div>
              </ProductForm>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </ProductFormHolder>
</withRouter(ProductFormHolder)>
`;
